File: frontend/src/types/book.ts
export type ReadingStatus = 'PENDING' | 'READING' | 'COMPLETED'

export interface Genre {
  id: string
  name: string
}

export interface Note {
  id: string
  user_book_id: string
  content: string
  page_number: number | null
  created_at: string
}

export interface Book {
  id: string
  title: string
  author: string
  description: string | null
  thumbnail: string | null
  pages: number | null
  published_date: string | null
  google_books_id: string | null
  created_at: string
  genres: Genre[]
}

export interface UserBook {
  id: string
  user_id: string
  book_id: string
  book: Book
  status: ReadingStatus
  current_page: number | null
  started_at: string | null
  finished_at: string | null
  rating: number | null
  notes: Note[]
  created_at: string
}

export interface CreateNoteRequest {
  content: string
  page_number?: number | null
}

export interface CreateBookRequest {
  title: string
  author: string
  description?: string
  thumbnail?: string
  pages?: number
  published_date?: string
  google_books_id?: string
  genre_ids?: string[]
}

export interface UpdateBookRequest {
  status?: ReadingStatus
  current_page?: number
  rating?: number
  started_at?: string
  finished_at?: string
  genre_ids?: string[]
}

File: backend/app/schemas/book.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime
from typing import Optional

from app.schemas.genre import GenreOut
from app.schemas.note import NoteOut


class BookOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    author: str
    description: Optional[str] = None
    thumbnail: Optional[str] = None
    pages: Optional[int] = None
    published_date: Optional[str] = None
    google_books_id: Optional[str] = None
    created_at: datetime
    genres: list[GenreOut] = []


class UserBookOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: str
    book_id: str
    book: BookOut
    status: str
    current_page: Optional[int] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    rating: Optional[float] = None
    notes: list[NoteOut] = []
    created_at: datetime


class CreateBookRequest(BaseModel):
    title: str
    author: str
    description: Optional[str] = None
    thumbnail: Optional[str] = None
    pages: Optional[int] = None
    published_date: Optional[str] = None
    google_books_id: Optional[str] = None
    genre_ids: Optional[list[str]] = None


class UpdateBookRequest(BaseModel):
    status: Optional[str] = None
    current_page: Optional[int] = None
    rating: Optional[float] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    genre_ids: Optional[list[str]] = None
